factcast-store/src/main/java/org/factcast/store/internal/tail/PGTailIndexManagerImpl.java

```diff
@@ -111,49 +93,111 @@ void createNewTail() {
+            StoreMetrics.VALUE.TAIL_INDICES, Tags.of(Tag.of(STATE, STATE_INVALID), maintenance))
+        .record(invalidIndexes.size());
+  }
+
+  private void processExistingIndices(JdbcTemplate jdbc) {
+    var indexesOrderedByTimeWithValidityFlag = getTailIndices(jdbc);
+    removeOldestValidIndices(jdbc, indexesOrderedByTimeWithValidityFlag);
+    removeInvalidIndices(jdbc, indexesOrderedByTimeWithValidityFlag);
+  }
+
+  @NonNull
+  private List<String> getValidIndices(List<Map<String, Object>> indexesWithValidityFlag) {
+    return new ArrayList<>(
+        indexesWithValidityFlag.stream()
+            .filter(r -> r.get(VALID_COLUMN).equals(IS_VALID))
+            .map(r -> r.get(INDEX_NAME_COLUMN).toString())
+            .toList());
   }
 
-  private void removeNonRecentInvalidIndices(List<Map<String, Object>> indexesWithValidityFlag) {
-    indexesWithValidityFlag.stream()
+  private @NonNull List<String> getInvalidIndices(
+      List<Map<String, Object>> indexesWithValidityFlag) {
+    return indexesWithValidityFlag.stream()
         .filter(r -> r.get(VALID_COLUMN).equals(IS_INVALID))
         .map(r -> r.get(INDEX_NAME_COLUMN).toString())
-        .filter(this::isNotRecent)
-        .forEach(this::removeIndex);
+        .toList();
   }
 
   @VisibleForTesting
-  void removeIndex(@NonNull String indexName) {
-    jdbc.update(PgConstants.dropTailIndex(indexName));
+  boolean timeToCreateANewTail(JdbcTemplate jdbc) {
+    var indexesOrderedByTimeWithValidityFlag = getTailIndices(jdbc);
+    var indices = getValidIndices(indexesOrderedByTimeWithValidityFlag);
+
+    if (indices.isEmpty()) {
+      return true;
+    }
+
+    var newestIndex = indices.get(0);
+
+    return exceedsMinimumTailAge(newestIndex);
+  }
+
+  @VisibleForTesting
+  protected boolean anyIndexOperationInProgress(JdbcTemplate jdbc) {
+    final var operations = jdbc.queryForList(INDEX_OPERATIONS_IN_PROGRESS);
+
+    if (operations.isEmpty()) {
+      return false;
+    }
+
+    log.debug("Index operations in progress: {}", operations);
+
+    return true;
+  }
+
+  @VisibleForTesting
+  protected void removeOldestValidIndices(
+      JdbcTemplate jdbc, List<Map<String, Object>> indexesWithValidityFlag) {
+    var validIndexes = getValidIndices(indexesWithValidityFlag);
+
+    while (validIndexes.size() > props.getTailGenerationsToKeep()) {
+      // Oldest is last in list as order is descending by name.
+      removeIndex(jdbc, validIndexes.remove(validIndexes.size() - 1));
+    }
+  }
+
+  private static @NonNull List<Map<String, Object>> getTailIndices(JdbcTemplate jdbc) {
+    return jdbc.queryForList(LIST_FACT_INDEXES_WITH_VALIDATION);
+  }
+
+  private void removeInvalidIndices(
+      JdbcTemplate jdbc, List<Map<String, Object>> indexesWithValidityFlag) {
+    getInvalidIndices(indexesWithValidityFlag).forEach(i -> removeIndex(jdbc, i));
+  }
+
+  @VisibleForTesting
+  void removeIndex(@NonNull JdbcTemplate jdbc, @NonNull String indexName) {
```

Comment:
should be have a precondition check that the indexname is indeed a tail index ?

Reply:
added :)